tree-sitter/src/TreeSitter/Unmarshal.hs

```diff
@@ -196,6 +201,10 @@ instance SymbolMatching f => SymbolMatching (Rec1 f) where
   symbolMatch _ = symbolMatch (Proxy @f)
   showFailure _ = showFailure (Proxy @f)
 
+instance (KnownNat n, KnownSymbol sym) => SymbolMatching (Token sym n) where
+  symbolMatch _ node = nodeSymbol node == fromIntegral (natVal (Proxy @n))
+  showFailure _ _ = "expected " ++ symbolVal (Proxy @sym)
```

Comment:
These definitions are slightly magical, but well-supported (the GHC generics metadata nodes employ the same tricks). `KnownSymbol` allows you to say that if a type-level string, like the `"+"` in the type `Token "+" 5`, is statically known, then you can recover it as a `String` value at runtime using that fact. `KnownNat` does the same thing for integers.

Thus, we can use the type-level information attached to the `Token` to both check whether we can match a given node, and to report an error if we can’t.